h5py/h5d.pyx

```diff
@@ -208,7 +208,7 @@ cdef class DatasetID(ObjectID):
         cdef int oldflags
 
         if mtype is None:
-            mtype = py_create(arr_obj.dtype)
+            mtype = arr_obj.id.get_type()
```

Comment:
I think this is the line that's causing the test failure - `arr_obj` is (or can be) a Numpy array.

I think we'd need a separate branch here, something like:

```python
ftype = self.get_type()
if isinstance(ftype, TypeOpaqueID) and ftype.get_tag() == b'NUMPY:' + arr_obj.dtype.descr[0][1].encode():
    mtype = ftype
```

(Of course, if `mtype` is passed, that should still override this check)

Reply:
This didn't work. I opted for `isinstance(ftype, TypeID)` instead, but there were still test failures... I'd like to avoid introducing special cases.

Reply:
Were they different test failures?

In any case, `arr_obj` being a numpy array is not a special case - that's the normal case that the function needs to handle.